### PR TITLE
MAINT: Remove malloc(0) calls in linalg and pocketfft

### PR DESCRIPTION
**Pull request by seberg**
This is inspired by Steven Pitmans start in gh-17961 and pulls out the fixes around `malloc(0)` calls (but in a different way.

I took the liberty of also adding at least some attempt at correct error handling for the umath code.

The precise way of avoiding 0s could be different in places...

### CODE REVIEW

File: numpy/linalg/umath_linalg.cpp [hunk 1]
```diff
@@ -1193,7 +1201,7 @@ det(char **args,
            void *NPY_UNUSED(func))
 {
     fortran_int m;
-    npy_uint8 *tmp_buff;
+    char *tmp_buff;
```
Comment (charris) [inline numpy/linalg/umath_linalg.cpp:1204]: See above.

**Comment by charris** [comment]
I'm wary of the use of `char` instead of `npy_uint8`. We have run into problems with it before, in particular, `char` is signed on Intel, unsigned on ARM.

**Comment by charris** [comment]
Apart the the `char` bit, this LGTM.

**Comment by seberg** [comment]
I prefer char for buffers that get cast, but happy to undo (This always gets cast).  I guess sizeof might be an argument, but we are probably broken anyway where, so I am not sure if there are more.

**Comment by charris** [comment]
> I prefer char for buffers that get cast

As long as the cast is forced and no arithmetic is done, that should be OK. The problems arise when one wants to use `char` as an integer. The original problem was in the einsum code where `char` was used as a small integer and needed to be able to represent -1.

**Comment by charris** [comment]
Thanks Sebastian.

**Pull request merged**